### PR TITLE
examples: Add taproot PSBT example workflow

### PR DESCRIPTION
**Pull request by dunxen**
Will address #893.

Currently includes a BIP86 example (no spendable script path)

Working on script path and key path spending when both are possible spending paths.

### CODE REVIEW

**Comment by apoelstra** [comment]
Looks like a great start!

**Comment by tcharding** [comment]
Looks good man, I'll be watching this one closely (your code and reviewers comments) so I can work out how to do https://github.com/rust-bitcoin/rust-bitcoin/pull/957. If you want to have input on that PR, that would be awesome too.

**Comment by sanket1729** [comment]
This looks great. Thanks for the detailed instructions, I will try out them to see if I can re-produce the results.

**Comment by dunxen** [comment]
Planning to get back to cleaning up the script path spending today and will try get it up this weekend. I just want to structure it properly otherwise it will be a bit messy to follow. It also requires instructions on actually constructing the output which is trickier IMO :)

**Comment by tcharding** [comment]
Bro, this is sick!

File: examples/taproot-psbt.rs
Comment (tcharding) [inline examples/taproot-psbt.rs:86]: This is stale now, right we have the println statement below that serves as documentation.

File: examples/taproot-psbt.rs
Comment (tcharding) [inline examples/taproot-psbt.rs:84]: Does this mean there is going to be an example 3 added or is 2.1 and 2.2 counted as two?

Comment (dunxen) [reply examples/taproot-psbt.rs:84]: Yeah, I was only going to have those 3 but it seems a little confusing so will renumber. I'd also like to show a larger taptree example but not sure if we need to worry about that too much?

**Comment by dunxen** [comment]
BDK's `sign_psbt_schnorr()` has mostly what we need for taproot support in the PSBT signing API PR @tcharding :)
I can add onto that one in a follow-up PR once you get it through. I'll review a bit of that today again too.

**Comment by tcharding** [comment]
I attempted to run through the example but got stuck on `Step 3`, I couldn't get bitcoin-cli to get a new taproot address.
```
bt -rpcwallet=benefactor getnewaddress '' bech32m
error code: -12
error message:
Error: No bech32m addresses available.
```
I tried creating wallets with a bunch of different options to no avail. I tried importing an xpriv also to no avail. The closest I got was I did manage to import an xpub and create a taproot address. Any ideas?

**Comment by dunxen** [comment]
> I attempted to run through the example but got stuck on `Step 3`, I couldn't get bitcoin-cli to get a new taproot address.
> 
> ```
> bt -rpcwallet=benefactor getnewaddress '' bech32m
> error code: -12
> error message:
> Error: No bech32m addresses available.
> ```
> 
> I tried creating wallets with a bunch of different options to no avail. I tried importing an xpriv also to no avail. The closest I got was I did manage to import an xpub and create a taproot address. Any ideas?

Thanks for trying to run it! I think I assumed these steps from a v23.0 perspective where descriptor wallets are default. I've updated it to indicate we need descriptor wallets. I _think_ that's the issue you're having if I'm not mistaken?

**Comment by tcharding** [comment]
oooh, I was using 22.0 assuming Taproot was all that's required. Serves me right for not updating :( Will play some more with it, cheers mate.

**Comment by tcharding** [comment]
I don't think its descriptors being default that is the problem because I was passing `descriptors=true` to `createwallet`. I started building bitcoind to get v23 but thought better of it. We should support the latest released version of bitcoind I rekon. I'll keep playing with v22 and see if I can work it out. Cheers.

File: examples/taproot-psbt.rs
Comment (tcharding) [inline examples/taproot-psbt.rs:9]: Whitespace on this line.

File: examples/taproot-psbt.rs
Comment (tcharding) [inline examples/taproot-psbt.rs:10]: I couldn't get this to work with v22 but v23 has been released now so I'd suggest just stating that the example docs are for Bitcoin Core v23. If you do that you can remove the `descriptors=true` from this and below.

File: examples/taproot-psbt.rs
Comment (tcharding) [inline examples/taproot-psbt.rs:34]: Perhaps add a comment saying "this is used for example 1", and likewise for the other two utxos.

File: examples/taproot-psbt.rs
Comment (tcharding) [inline examples/taproot-psbt.rs:41]:
```suggestion
    derivation_path: "m/86'/1'/0'/0/0", // The derivation path for the key that can spend this UTXO (from descriptor in step 4), you shouldn't need to change this.
```

File: examples/taproot-psbt.rs
Comment (tcharding) [inline examples/taproot-psbt.rs:52]: Perhaps use consts for the values that users are not expected to change?

File: examples/taproot-psbt.rs
Comment (tcharding) [inline examples/taproot-psbt.rs:78]: Can you group by module please, like we do elsewhere in `rust-bitcoin`.

File: examples/taproot-psbt.rs
Comment (tcharding) [inline examples/taproot-psbt.rs:82]: No need for this anymore, it is in the `psbt` module now.

File: examples/taproot-psbt.rs
Comment (tcharding) [inline examples/taproot-psbt.rs:93]: Perhaps set these in consts up the top like the utxos so there is nothing that needs editing in main?

File: examples/taproot-psbt.rs
Comment (tcharding) [inline examples/taproot-psbt.rs:40]:
```suggestion
    amount_in_sats: 50 * COIN_VALUE, // Change '50' to the amount in your utxo (must be at least 1).
```

File: examples/taproot-psbt.rs
Comment (tcharding) [inline examples/taproot-psbt.rs:160]:
```suggestion
    let (tx, psbt) = benefactor.create_inheritance_funding_tx(1200, UTXO_2)?;
```

File: examples/taproot-psbt.rs
Comment (tcharding) [inline examples/taproot-psbt.rs:170]:
```suggestion
    let spending_tx = beneficiary.spend_inheritance(psbt, 1200, &to_address, 1000)?;
    let spending_tx_hex = encode::serialize_hex(&spending_tx);
```

Note I used 1200 instead of 1000 otherwise one cannot verify the locktime if example two has already been done.

File: examples/taproot-psbt.rs
Comment (tcharding) [inline examples/taproot-psbt.rs:173]: If you change the locktime then this will need updating too.

File: examples/taproot-psbt.rs
Comment (tcharding) [inline examples/taproot-psbt.rs:418]: Is this still a todo?

File: examples/taproot-psbt.rs
Comment (tcharding) [inline examples/taproot-psbt.rs:447]:
```suggestion
        psbt.outputs = vec![Output::default()];
```

**Review by tcharding — COMMENTED** [review]
Nice one man, I got it all running using Bitcoin Core v23. I made a few minor changes to the third example so that all three examples could be worked through. Gave a few docs suggestions, use them as you wish. Gave a couple minor code suggestions but I did not review the exact logic of the two wallet types, simply because I'm not knowledgeable to do so.

I can confirm that I worked through all three examples.

File: examples/taproot-psbt.rs
Comment (tcharding) [reply examples/taproot-psbt.rs:78]: To clarify
```
psbt::{serialize::Serialize, Input, PartiallySignedTransaction, PsbtSighashType, Output, self},
```
becomes
```
psbt::serialize::Serialize;
psbt::{Input, PartiallySignedTransaction, PsbtSighashType, Output, self},
```

File: examples/taproot-psbt.rs
Comment (tcharding) [reply examples/taproot-psbt.rs:170]: We could use a const for the fee as well so as not to confuse it with locktime?

File: examples/taproot-psbt.rs
Comment (tcharding) [inline examples/taproot-psbt.rs:79]:
```suggestion
use bitcoin::hashes::Hash;
```

File: examples/taproot-psbt.rs
Comment (dunxen) [reply examples/taproot-psbt.rs:418]: Yes, indeed! Sorry, just getting back to this today :)

File: examples/taproot-psbt.rs
Comment (dunxen) [reply examples/taproot-psbt.rs:447]: Of course :man_facepalming: :sweat_smile:

**Comment by dunxen** [comment]
> @dunxen  Is this ready for review? If you have some more changes to make, feel free to tag me for review once you think this is ready. No rush :) 

Been working on the last bits today and adding some changes from  previous review.

Thanks, I will definitely tag you when ready! 🙏

**Comment by dunxen** [comment]
Just needed a rebase, but will limit the squashing/rebasing from now when I've promoted it from draft. There's probably still a bit of cleanup and dedup left.

**Comment by tcharding** [comment]
Note, if #1079 goes in before this we will need to add a clippy call to `test.sh`.

**Review by tcharding — APPROVED** [review]
ACK 8df9a0b31cfca420c8949d619905faee789a84bb

**Comment by dunxen** [comment]
Rebasing and getting this ready for a final review. Apologies for the hold up.

**Comment by tcharding** [comment]
To fix CI use:
```rust
"\nYou should now be able to broadcast the following transaction: \n\n{}", tx_hex_string
```

**Comment by tcharding** [comment]
Looks good to me bro, I had a bit of a fiddle with the imports, mainly because they displayed how untidy our API is (cleaning it up is one thing I'm working on right now). No need to change but if you are interested this is what I did
```diff
diff --git a/bitcoin/examples/taproot-psbt.rs b/bitcoin/examples/taproot-psbt.rs
index e43f7546..043d8f86 100644
--- a/bitcoin/examples/taproot-psbt.rs
+++ b/bitcoin/examples/taproot-psbt.rs
@@ -78,30 +78,26 @@ const UTXO_3: P2trUtxo = P2trUtxo {
 use std::collections::BTreeMap;
 use std::str::FromStr;
 
-use bitcoin::absolute::PackedLockTime;
 use bitcoin::consensus::encode;
 use bitcoin::constants::COIN_VALUE;
 use bitcoin::hashes::hex::FromHex;
-use bitcoin::locktime::absolute;
+use bitcoin::hashes::Hash;
 use bitcoin::opcodes::all::{OP_CHECKSIG, OP_CLTV, OP_DROP};
 use bitcoin::psbt::serialize::Serialize;
-use bitcoin::psbt::Output;
+use bitcoin::psbt::{self, Input, Output, Psbt, PsbtSighashType};
 use bitcoin::schnorr::TapTweak;
-use bitcoin::secp256k1::Secp256k1;
+use bitcoin::secp256k1::{Message, Secp256k1};
 use bitcoin::util::bip32::{
     ChildNumber, DerivationPath, ExtendedPrivKey, ExtendedPubKey, Fingerprint,
 };
-use bitcoin::util::psbt::{self, Input, Psbt, PsbtSighashType};
 use bitcoin::util::sighash;
 use bitcoin::util::taproot::{
     LeafVersion, TapLeafHash, TapSighashHash, TaprootBuilder, TaprootSpendInfo,
 };
 use bitcoin::{
-    script, Address, Amount, OutPoint, SchnorrSig, SchnorrSighashType, Script, SighashCache,
-    Transaction, TxIn, TxOut, Txid, Witness, XOnlyPublicKey,
+    absolute, script, Address, Amount, OutPoint, SchnorrSig, SchnorrSighashType, Script,
+    SighashCache, Transaction, TxIn, TxOut, Txid, Witness, XOnlyPublicKey,
 };
-use bitcoin_hashes::Hash;
-use secp256k1::Message;
 
 fn main() -> Result<(), Box<dyn std::error::Error>> {
     let secp = Secp256k1::new();
@@ -133,7 +129,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
         ],
     )?);
     println!(
-        "\nYou should now be able to broadcast the following transaction: \n\n{tx_hex_string}",
+        "\nYou should now be able to broadcast the following transaction: \n\n{}", tx_hex_string
     );
 
     println!("\nEND EXAMPLE 1\n");
@@ -637,7 +633,7 @@ impl BeneficiaryWallet {
         let input_value = psbt.inputs[0].witness_utxo.as_ref().unwrap().value;
```

**Review by tcharding — APPROVED** [review]
ACK c05188a878501e8ff953f88b47897431e4385536

**Comment by dunxen** [comment]
> To fix CI use:
> 
> ```rust
> "\nYou should now be able to broadcast the following transaction: \n\n{}", tx_hex_string
> ```

Oh yeah. MSRV things

**Comment by dunxen** [comment]
We can clean this up nicely when we extend the PSBT input signing API for schnorr/taproot :)

**Review by tcharding — APPROVED** [review]
ACK 1a89d5230c0db044d17b21fccedf18bad25c6c08

**Comment by tcharding** [comment]
Can we get a second ACK on this one please? The PR only adds code to `bitcoin/examples` so there is no risk of merging without deep security review. All the other psbt example and signing improvements have merged now.

**Review by apoelstra — APPROVED** [review]
ACK 1a89d5230c0db044d17b21fccedf18bad25c6c08

**Comment by tcharding** [comment]
BOOM!

**Pull request merged**
